From edfd5887f77a65779e8215cc89962ae84293e1bd Mon Sep 17 00:00:00 2001
From: Lily Zhang <r58066@freescale.com>
Date: Sat, 14 May 2011 15:46:14 +0800
Subject: [PATCH 962/998] ENGR00143430 mx53 ard: remove DDR operations in suspend and resume

The system can not resume when running the following test case in
mx53 ARD RevB board:
/unit_tests/rtcwakeup.out -m mem -s 2
After removing DDR operation in suspend/resume, the issue disappears.

Signed-off-by: Lily Zhang <r58066@freescale.com>
---
 arch/arm/mach-mx5/pm.c      |    4 ++++
 arch/arm/mach-mx5/suspend.S |    5 +++++
 2 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx5/pm.c b/arch/arm/mach-mx5/pm.c
index a2e35d1..894484c 100644
--- a/arch/arm/mach-mx5/pm.c
+++ b/arch/arm/mach-mx5/pm.c
@@ -249,6 +249,7 @@ static struct platform_driver mx5_pm_driver = {
 
 #define SUSPEND_ID_MX51 1
 #define SUSPEND_ID_MX53 3
+#define SUSPEND_ID_NONE 4
 static int __init pm_init(void)
 {
 	unsigned long iram_paddr;
@@ -270,6 +271,9 @@ static int __init pm_init(void)
 	if (cpu_is_mx51() || cpu_is_mx53()) {
 		suspend_param1 =
 			cpu_is_mx51() ? (void *)SUSPEND_ID_MX51:(void *)SUSPEND_ID_MX53;
+		/* for mx53 ARD, doesn't operate DDR in suspend */
+		if (machine_is_mx53_ard())
+			suspend_param1 = (void *)SUSPEND_ID_NONE;
 		memcpy(suspend_iram_base, cpu_do_suspend_workaround,
 				SZ_4K);
 	} else if (cpu_is_mx50()) {
diff --git a/arch/arm/mach-mx5/suspend.S b/arch/arm/mach-mx5/suspend.S
index b45d72f..1203c31 100644
--- a/arch/arm/mach-mx5/suspend.S
+++ b/arch/arm/mach-mx5/suspend.S
@@ -136,6 +136,9 @@ FinishedClean:
     bic     r0, r0, #ARM_AUXCR_L2EN @ Disable L2 cache
     mcr     p15, 0, r0, c1, c0, 1   @ Update aux control reg
 
+    /* Do nothing for DDR */
+    cmp     r6, #4
+    beq     mx5x_wfi
 	/*Set the DDR drive strength to low */
     cmp     r6, #1
     bne     mx53_reduce_ddr_drive_strength
@@ -183,6 +186,8 @@ mx5x_wfi:
     mov r0,#0x0
     .long     0xe320f003              @ Opcode for WFI
 
+    cmp     r6, #4
+    beq     mx5x_post_wfi
 	/*Set the DDR drive strength to max */
     cmp     r6, #1
     bne     mx53_restore_ddr_drive_strength
-- 
1.7.4.1

